src/ctx.rs

```diff
@@ -57,42 +81,146 @@ impl AppConfig {
             .unwrap_or(DEFAULT_TOKEN_EXPIRATION_TIME)
     }
 
-    pub(crate) fn get_proxy_route_by_inbound(&self, inbound: String) -> Option<&ProxyRoute> {
+    pub(crate) fn get_proxy_route_by_inbound(&self, inbound: &str) -> Option<&ProxyRoute> {
         let route_index = self.proxy_routes.iter().position(|r| {
-            r.inbound_route == inbound || r.inbound_route.to_owned() + "/" == inbound
+            r.inbound_route == inbound
+                || r.inbound_route == "/".to_owned() + inbound
+                || r.inbound_route.to_owned() + "/" == inbound
+                || "/".to_owned() + &*r.inbound_route == "/".to_owned() + inbound
         });
 
-        if let Some(index) = route_index {
-            return Some(&self.proxy_routes[index]);
+        route_index.map(|index| &self.proxy_routes[index])
+    }
+
+    /// Finds the best matching proxy route based on the provided URI's path and updates the URI by
+    /// removing the matched path segments while preserving the query parameters.
+    pub(crate) fn get_proxy_route_extracting_uri_inbound(
+        &self,
+        uri: &mut Uri,
+    ) -> GenericResult<Option<&ProxyRoute>> {
+        let path_segments: Vec<&str> = uri.path().split('/').filter(|s| !s.is_empty()).collect();
+
+        let mut best_match: Option<(&ProxyRoute, usize)> = None;
+
+        for r in &self.proxy_routes {
+            let route_segments: Vec<&str> = r
+                .inbound_route
+                .split('/')
+                .filter(|s| !s.is_empty())
+                .collect();
+            // Count the number of segments that match between the route and the path
+            let matched_segments = route_segments
+                .iter()
+                .zip(&path_segments)
+                .take_while(|(route_seg, path_seg)| route_seg == path_seg)
+                .count();
```

Comment:
Not a blocker but depending on the request and route configuration, this can cause lots of iterations (from parent to sub iterations).

I did something like this, which should be more sufficient:


```rs
fn find_route_from_url(incoming_url: &Url, routes: &[Route]) -> Option<Route> {
    let path = incoming_url.path();
    let mut matched_route: Option<&Route> = None;

    for route in routes {
        if path.starts_with(&route.inbound) {
            if let Some(r) = matched_route {
                if r.inbound.len() > route.inbound.len() {
                    continue;
                }
            }
            matched_route = Some(route);
        }
    }

    matched_route.cloned()
}

fn inbound_url_to_outbound_url(inbound_url: Url, route: &Route) -> Url {
    let inbound_url_str = inbound_url.to_string();
    let path = inbound_url_str
        .split_once(inbound_url.host_str().unwrap())
        .expect("inbound url must contain a host")
        .1;

    let path = path
        .strip_prefix(&route.inbound)
        .expect("Route doesn't match with the given inbound url.");

    let url = format!(
        "{}/{}",
        route.outbound.strip_suffix('/').unwrap_or(&route.outbound),
        path.strip_prefix('/').unwrap_or(&path)
    );

    Url::parse(&url).expect("valid url")
}
```


the test function:

```rs
#[test]
fn find_best_match() {
    fn test_helper(inbound_url: Url, routes: &[Route]) -> String {
        let route = find_route_from_url(&inbound_url, routes).unwrap();
        let result = inbound_url_to_outbound_url(inbound_url, &route);
        result.to_string()
    }

    let routes = vec![
        Route {
            inbound: "/".into(),
            outbound: "https://k.com".into(),
        },
        Route {
            inbound: "/demo".into(),
            outbound: "https://kmd.com".into(),
        },
        Route {
            inbound: "/demo/demo".into(),
            outbound: "https://komodo.com".into(),
        },
        Route {
            inbound: "/demo/demo/demo".into(),
            outbound: "https://kdf.com/komodo".into(),
        },
    ];

    let inbound = Url::parse("https://rproxy-app.service/hello/there").unwrap();
    let expected_outbound_url = "https://k.com/hello/there";
    let actual_outbound_url = test_helper(inbound, &routes);
    assert_eq!(actual_outbound_url, expected_outbound_url);

    let inbound = Url::parse("https://rproxy-app.service/demo/some/stuff/").unwrap();
    let expected_outbound_url = "https://kmd.com/some/stuff/";
    let actual_outbound_url = test_helper(inbound, &routes);
    assert_eq!(actual_outbound_url, expected_outbound_url);

    let inbound = Url::parse("https://rproxy-app.service/demo/demo/okay?q=123").unwrap();
    let expected_outbound_url = "https://komodo.com/okay?q=123";
    let actual_outbound_url = test_helper(inbound, &routes);
    assert_eq!(actual_outbound_url, expected_outbound_url);

    let inbound = Url::parse("https://rproxy-app.service/demo/demo/demo/1/2/3").unwrap();
    let expected_outbound_url = "https://kdf.com/komodo/1/2/3";
    let actual_outbound_url = test_helper(inbound, &routes);
    assert_eq!(actual_outbound_url, expected_outbound_url);
}
```

Reply:
> I did something like this, which should be more sufficient

in your example you use Url type from url crate. Do you suggest to add this dependency? Currently we use only Uri

Reply:
Updated in this commit https://github.com/KomodoPlatform/komodo-defi-proxy/pull/22/commits/e3d81fa408562697588947cfba18b2db14a62e86

moved `delete inbound from Uri logic` back to modify uri function, so get route just gets proxy route, modify function modifies Uri